consensus/tendermint/committee/default.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/clearmatics/autonity/consensus/tendermint/config"
+	"github.com/clearmatics/autonity/core/types"
+	"math"
+	"reflect"
+	"sort"
+	"sync"
+)
+
+var ErrEmptyCommitteeSet = errors.New("committee set can't be empty")
+
+type defaultSet struct {
+	members      types.Committee
+	policy       config.ProposerPolicy
+	lastProposer common.Address
+	selector     ProposalSelector
+
+	mu       sync.RWMutex                    // members doesn't need to be protected as it is read-only
+	proposer map[int64]types.CommitteeMember // cached computed values
+}
+
+func NewSet(members types.Committee, policy config.ProposerPolicy, lastProposer common.Address) (*defaultSet, error) {
+
+	if len(members) == 0 {
+		return nil, ErrEmptyCommitteeSet
+	}
+
+	committee := &defaultSet{}
+	committee.policy = policy
+	committee.members = members
+	committee.proposer = make(map[int64]types.CommitteeMember)
+	// sort validator
+	sort.Sort(committee.members)
+
+	switch policy {
+	case config.Sticky:
+		committee.selector = stickyProposer
+	case config.RoundRobin:
+		committee.selector = roundRobinProposer
+	default:
+		committee.selector = roundRobinProposer
+	}
+
+	committee.lastProposer = lastProposer
+	committee.proposer[0] = committee.selector(committee, lastProposer, 0)
+	return committee, nil
+}
+
+func copyMembers(members types.Committee) types.Committee {
+	membersCopy := make(types.Committee, len(members))
+	for i, val := range members {
+		membersCopy[i] = val
+	}
+	return membersCopy
+}
+
+func (set *defaultSet) Size() int {
+	return len(set.members)
+}
+
+func (set *defaultSet) Committee() types.Committee {
+	return copyMembers(set.members)
+}
+
+func (set *defaultSet) GetByIndex(i int) (types.CommitteeMember, error) {
+	if i < 0 || i >= len(set.members) {
+		return types.CommitteeMember{}, consensus.ErrCommitteeMemberNotFound
+	}
+	return set.members[i], nil
+}
+
+func (set *defaultSet) GetByAddress(addr common.Address) (int, types.CommitteeMember, error) {
+	for i, member := range set.members {
+		if addr == member.Address {
+			return i, member, nil
+		}
+	}
+	return -1, types.CommitteeMember{}, consensus.ErrCommitteeMemberNotFound
+}
+
+func (set *defaultSet) GetProposer(round int64) types.CommitteeMember {
+	set.mu.Lock()
+	defer set.mu.Unlock()
+	v, ok := set.proposer[round]
+	if !ok {
+		v = set.selector(set, set.lastProposer, round)
+		set.proposer[round] = v
+	}
+
+	return v
+}
+
+func (set *defaultSet) IsProposer(round int64, address common.Address) bool {
+	_, val, err := set.GetByAddress(address)
+	if err != nil {
+		return false
+	}
+	curProposer := set.GetProposer(round)
+	return reflect.DeepEqual(curProposer, val)
+}
+
+func (set *defaultSet) Copy() Set {
+	newSet, _ := NewSet(copyMembers(set.members), set.policy, set.lastProposer)
+	return newSet
+}
+
+func (set *defaultSet) F() int { return int(math.Ceil(float64(set.Size())/3)) - 1 }
+
+func (set *defaultSet) Quorum() int { return int(math.Ceil((2 * float64(set.Size())) / 3.)) }
```

Comment:
This quorum size calculation is wrong, since it can return exactly 2/3 of the set size where it should be returning a number > 2/3 of the set size.

For example given a set of size 6 it returns 4 instead of 5.

```suggestion
func (set *defaultSet) Quorum() int { return int(2.0/3.0*float64(set.Size()))+1 }
```

Reply:
This formula is correct,  `Q > 2N/3` is actually not always true, you can find the formal demonstration somewhere in this paper https://arxiv.org/pdf/1901.07160.pdf

if you want to convince yourself, for a network of 6 validators we  can only support 1 faulty node.

1.` N-f >= Q,` there must be at least a quorum of correct processes to ensure liveness, 6 - 1 = 5 >= 4 
2. `2 Q - N > f`, the intersection of two quorums must be greater than the number of faulty nodes.  2 x 4 - 6 = 2 > 1